### PR TITLE
Add endpoint to retrieve multiaddr of nw primary

### PR DESCRIPTION
**Pull request by arun-koshy**
Add endpoint to expose primary to primary interface. This endpoint will be used to publish multiaddr of the new primary to an announce protocol. 

Endpoint usage as explained by @huitseeker 
> Lifecycle of a NW MultiAddr
> 
> 1. (NW process : own MultiAddr) —- RPC endpoint ——> (Consensus process: own MultiAddr)
> 2. (Consensus process: own MultiAddr). ——- DHT / Announce —-> (DHT : registered MultiAddr for node PK)
> 3.  (DHT : registered MultiAddr for node PK) ——- DHT / Annouce —→ (x size of deployment) (DHT : registered MultiAddr for all nodes in NW)
> 4. (Consensus : registered MultiAddr for all nodes in NW) —— RPC Server ——> (NW process: registered MultAddr for all nodes in NW)
> 
> Step 1. is local RPC, 2. is DHT protocol write 3. is DHT listening (and subsumes a gossip boundary) 4. is local RPC.

### CODE REVIEW

File: primary/tests/integration_tests_configuration_api.rs
Comment (akichidis) [inline primary/tests/integration_tests_configuration_api.rs:186]: I believe you can benefit in this test by [Cluster](https://github.com/MystenLabs/narwhal/blob/main/test_utils/src/cluster.rs). It will allow you to easily spin up - even just one node - and perform this test without having to do this setup . A small addition might be needed so the cluster can run without consensus - I'll see whether I can have something quickly so you can use it here.

File: primary/src/grpc_server/configuration.rs
Comment (akichidis) [inline primary/src/grpc_server/configuration.rs:14]: Nit: I would just go with `primary_address` , we don't gain much by truncating

File: primary/tests/integration_tests_configuration_api.rs
Comment (akichidis) [inline primary/tests/integration_tests_configuration_api.rs:236]: The need to have a regex seems to me that the exact address is not known due to some dynamic assignment, but can't we derive the exact address from the `committee` and compare against an exact match?

File: primary/src/grpc_server/configuration.rs
Comment (akichidis) [inline primary/src/grpc_server/configuration.rs:20]: Since the `primary_addr` is derived from the `comittee` can't we just derive it from them when we initialise the struct, so we don't have the need to pass it separately?

**Review by akichidis — APPROVED** [review]
Looks good to me 👍, thanks @arun-koshy!  A few comments but approving to unblock

File: primary/tests/integration_tests_configuration_api.rs
Comment (akichidis) [reply primary/tests/integration_tests_configuration_api.rs:186]: @arun-koshy you can find the changes on [this PR](https://github.com/MystenLabs/narwhal/pull/599) . An example of how you could use the `Cluster` in this test can be found [here](https://github.com/MystenLabs/narwhal/pull/599/files#diff-23e4050e4099fea2c66148e23cfb98113529fa70502a78764b98494536b4c19dR39)

File: primary/src/grpc_server/configuration.rs
Comment (arun-koshy) [reply primary/src/grpc_server/configuration.rs:20]: We would need the name (public key) of the primary to be able to derive it from the committee so I could either pass the name or the primary address directly in. Maybe the name is better if we ever want to derive other information from the committee?

File: primary/tests/integration_tests_configuration_api.rs
Comment (arun-koshy) [reply primary/tests/integration_tests_configuration_api.rs:186]: This is awesome! Thanks Tasos 🤩

Comment (arun-koshy) [reply primary/tests/integration_tests_configuration_api.rs:186]: Will wait for your PR to merge, to properly utilize your changes.

File: primary/src/grpc_server/configuration.rs
Comment (huitseeker) [inline primary/src/grpc_server/configuration.rs:161]: Since this always succeed, does it need to return a `Result`?

File: test_utils/src/cluster.rs
Comment (huitseeker) [inline test_utils/src/cluster.rs:648]: I don't think I see any risk in letting this be accessible with the internal consensus enabled. Do you?

**Review by huitseeker — COMMENTED** [review]
This looks like it's getting in pretty good shape!

File: primary/src/grpc_server/configuration.rs
Comment (arun-koshy) [reply primary/src/grpc_server/configuration.rs:161]: tonic requires `Result` here

```
 note: expected fn pointer `fn(&'life0 NarwhalConfiguration, tonic::Request<_>) -> Pin<Box<(dyn futures::Future<Output = Result<tonic::Response<GetPrimaryAddressResponse>, Status>> + std::marker::Send + 'async_trait)>>`
 ```

File: test_utils/src/cluster.rs
Comment (arun-koshy) [reply test_utils/src/cluster.rs:648]: It won't be able to connect as the grpc server only starts when internal consensus is disabled [here](https://github.com/MystenLabs/narwhal/blob/main/primary/src/primary.rs#L331-L346).

File: primary/src/grpc_server/configuration.rs
Comment (arun-koshy) [reply primary/src/grpc_server/configuration.rs]: @akichidis I changed this to pass in the name instead of primary address, if you had something else in mind let me know and I will change it in a future PR.

**Pull request merged**
